cli_tools/common/distro/distro.go

```diff
@@ -33,155 +36,216 @@ const (
 // Release encapsulates product and version information
 // about the operating systems we support.
 type Release interface {
+	// ImportCompatible returns whether a distro considers two releases
+	// compatible. As an example, different minor versions in Ubuntu are typically
+	// not compatible, while different minor versions in centOS are.
 	ImportCompatible(other Release) bool
-	AsGcloudArg() string
-}
-
-type release struct {
-	distro  string
-	variant string
-	major   string
-	minor   string
-}
 
-// AsGcloudArg returns a string representation using the format
-// that we use with the `--os` flag.
-func (l release) AsGcloudArg() string {
-	switch l.distro {
-	case ubuntu:
-		return fmt.Sprintf("%s-%s%s", l.distro, l.major, l.minor)
-	case sles:
-		if l.variant != "" {
-			return fmt.Sprintf("%s-%s-%s", l.distro, l.variant, l.major)
-		}
-		return fmt.Sprintf("%s-%s", l.distro, l.major)
-	default:
-		return fmt.Sprintf("%s-%s", l.distro, l.major)
-	}
-}
-
-// ImportCompatible reports whether two releases are compatible for import.
-func (l release) ImportCompatible(other Release) bool {
-	realOther, ok := other.(release)
-	if !ok {
-		return false
-	}
-
-	if l.distro == ubuntu {
-		return l.distro == realOther.distro &&
-			l.major == realOther.major &&
-			l.minor == realOther.minor
-	}
-
-	if l.distro == sles {
-		return l.distro == realOther.distro &&
-			l.variant == realOther.variant &&
-			l.major == realOther.major
-	}
-
-	return l.distro == realOther.distro &&
-		l.major == realOther.major
+	// AsGcloudArg returns a hyphenated identifier, as used in the `--os` flag
+	// of `gcloud compute images import`. It is intended for
+	// showing to users in a help message. Notably, it lacks support for
+	// the -byol suffix, as that is not currently modeled.
+	AsGcloudArg() string
 }
 
-// FromLibguestfs initializes a DistroAndVersion from the fields returned by libguestfs's
-// inspection routines.
+// FromComponents interprets the (distro, major, minor) tuple and returns
+// a Release if the arguments are syntactically correct, and represent a
+// release that we *may* support. The caller is responsible for verifying
+// whether a translator is available elsewhere in the system.
 //
-// http://libguestfs.org/guestfs.3.html#guestfs_inspect_get_distro
-func FromLibguestfs(distro string, major string, minor string) (r Release, e error) {
+// Prefer this method for generating a new Release object. It is lenient, and
```

Comment:
Prefer to what?

Reply:
Hmm, good point. Originally I intended it to warn against using the distro-specific constructors, but those aren't exported, so that point is moot. Maybe just remove these last two lines?

Reply:
removed